### PR TITLE
Connect: Add prerequisites for gracefully handling expired db proxy certs

### PR DESCRIPTION
**Pull request by ravicious**
The first three commits are small cleanups around gateways. The next four commits are prerequisites for reissuing certs in Connect's db proxies (#17950).

## Rationale

#16958 added a middleware implementation for `alpn.LocalProxy`. This way if a client tries to make a connection through the proxy but the db cert has expired, `tsh proxy db` is able to ask the user to log in again and reissue the cert. We want to implement a similar behavior in Connect.

That middleware from #16958 also supports per-session MFA. In Connect however we need to just detect if the cert has expired. In the future, when we get around to adding per-session MFA support to Connect, we plan to refactor the existing middleware from #16958 so that it can be reused between tsh and Connect.

This PR provides the plumbing required by #17950. It adds a simpler implementation of LocalProxy middleware which just checks the cert expiry and fires the provided callback if the cert has expired. In #17950, I'm going to implement the rest of the logic: within the callback I'm going to ask the user to relogin and then reissue the cert.

Each of the four commits explains why a particular change is needed.

### CODE REVIEW

File: lib/teleterm/gateway/gateway.go
Comment (ravicious) [inline lib/teleterm/gateway/gateway.go:121]: It's kind of an awkward way to initialize the middleware and add the callback later. However, I want to be able to pass `gateway.Gateway` as the argument so that's why I first instantiate the middleware, then the gateway and then set the callback.

**Comment by ravicious** [comment]
Wow, I feel like I won the lottery with the auto review requests here. :d I wanted to request a review from Gavin anyway as he worked on #16958.

File: lib/teleterm/gateway/gateway.go
Comment (ravicious) [inline lib/teleterm/gateway/gateway.go:246]: Now that I'm looking at this, I'm not sure if this mutex here is needed. `alpn.LocalProxy` processes connections serially, so there's no risk of two separate goroutines trying to set the cert.

Comment (GavinFrazar) [reply lib/teleterm/gateway/gateway.go:246]: While reviewing this code I realized I needed a mutex in the local proxy still.

See: #18278

The window for a data race is extremely small, but we do call GetCerts to copy the certs into the TLS config when we dial upstream.

I originally had a mutex in there, but then we made the middleware a blocking call and I thought it was safe to remove.

That being said, once that PR merges I don't think you need a mutex here as the local proxy will guard its own certs with a mutex.

File: lib/teleterm/gateway/gateway_test.go
Comment (GavinFrazar) [inline lib/teleterm/gateway/gateway_test.go:98]: I'm unfamiliar with this test code - is there any reason you need a specific port? Can we just listen with ":0" and let the OS allocate an available port?

File: lib/teleterm/gateway/local_proxy_middleware.go
Comment (GavinFrazar) [inline lib/teleterm/gateway/local_proxy_middleware.go:72]: I added the `OnStart` function to the interface because we wanted to ensure that the user is prompted for credentials if needed when they run `tsh proxy db --tunnel` instead of waiting for a new connection to come in. Reading your comments it looks like that isn't a concern for teleterm, but I just wanted to clarify why the function exists. As long as teleterm prompts for credentials before starting the local proxy this is fine.

File: lib/teleterm/gateway/local_proxy_middleware_test.go
Comment (GavinFrazar) [inline lib/teleterm/gateway/local_proxy_middleware_test.go:44]: 😉

File: lib/teleterm/clusters/cluster_gateways.go
Comment (GavinFrazar) [inline lib/teleterm/clusters/cluster_gateways.go:40]: I don't think this is assigned anywhere?

**Review by GavinFrazar — CHANGES_REQUESTED** [review]
Mostly looks good but I think the OnExpiredCert func is never assigned. Your test mocked it so that's why it passes there. There's an integration test, maybe you could add something similar for the teleterm middleware: https://github.com/gravitational/teleport/blob/7676f3819937cdaeeaf19aedd57ca3db796c962e/integration/proxy/proxy_test.go#L759

edit: if you plan to add that in a separate PR I think that's fine, just wanted to make sure this was intentional. I can approve if that's the case, my other comments are minor things.

File: lib/teleterm/gateway/gateway_test.go
Comment (ravicious) [reply lib/teleterm/gateway/gateway_test.go:98]: `gateway.NewWithLocalPort` is used in Connect when the user wants to change the port number of the proxy. Take a look at the screenshot that I posted above, there's a field for the port number next to the database name field.

Of course, in tests we don't want to listen on a specific port. But I still want to test that when the user asks the app to change the port to 12345, then the port is actually changed to that number. So instead of using `net.Listen` in `gateway.Gateway`, I added `TCPPortAllocator` interface. In tests I use mock listeners which obtain a random port and `mockListener.Addr()` reports a fake address, in this case it'd be `localhost:12345`.

https://github.com/gravitational/teleport/blob/7676f3819937cdaeeaf19aedd57ca3db796c962e/lib/teleterm/gatewaytest/helpers.go#L95-L131

File: lib/teleterm/gateway/config.go
Comment (ravicious) [reply lib/teleterm/gateway/config.go:77]: The `OnExpiredCert` field of `Config` has a comment a few lines above. In those situations, do I just copy the comment from `OnExpiredCert` and paste it here? 🙊

File: lib/teleterm/clusters/cluster_gateways.go
Comment (GavinFrazar) [reply lib/teleterm/clusters/cluster_gateways.go:54]: Hmm, I was not aware of this - I just tested this and my postgres db cert omits the schema name extension (1.3.9999.2.4).  However, if I try to connect it uses my --db-user as the --db-name. Not sure where that's happening

File: lib/teleterm/gateway/config.go
Comment (GavinFrazar) [reply lib/teleterm/gateway/config.go:77]: I would just copy it since the type is exported

**Comment by ravicious** [comment]
Adding the mutex to gateway.Gateway turned out to be too complex for this PR, and also not necessary for the new feature it adds (see https://github.com/gravitational/teleport/pull/18259#discussion_r1016883107).

Still, it's worthwhile to add it but I'm going to move this to a completely new PR since it requires some bigger changes and I don't want to pollute this one even more.

**Comment by ravicious** [comment]
@GavinFrazar @smallinsky ping

The last commit makes the middleware work with changes from #18278.

File: lib/teleterm/clusters/cluster_gateways.go
Comment (smallinsky) [reply lib/teleterm/clusters/cluster_gateways.go:54]: In case of Postgres and Mongo databaseName is part part of Wire database protocol: For instance: 
https://github.com/gravitational/teleport/blob/1b0a083b9df37a9390eaaf57c3308e4b2a1783b4/lib/srv/db/postgres/engine.go#L177

The `--db-name` sets the value in connection string that is used my pgwire protocol and propagated in StartupMessage so event if the `RouteToDatabase.Database` is not set we are obtaining database name that was set by --db-name from the database wire protocol.

File: lib/teleterm/gateway/local_proxy_middleware_test.go
Comment (smallinsky) [inline lib/teleterm/gateway/local_proxy_middleware_test.go:104]: I think that we should call Close for read conn and write conn ends of pipe to prevent resource leak.

File: lib/teleterm/gateway/local_proxy_middleware_test.go
Comment (smallinsky) [inline lib/teleterm/gateway/local_proxy_middleware_test.go:106]: the `conn` seems to be unused in https://github.com/gravitational/teleport/pull/18259/files#diff-ec19972f09c27da3fcf2a7ed6f593473e0455bd5b1d4f888512e4338e6820977R40
Can we just pass a `nil` ?

File: lib/teleterm/gateway/local_proxy_middleware.go
Comment (smallinsky) [inline lib/teleterm/gateway/local_proxy_middleware.go:43]: The CheckDBCerts seems to also returns diffrents errors apart from `x509.CertificateInvalidError` like: `certificate subject is for user` Should we call onExpireCert login in case of "certificate subject is for user" error ?

**Review by smallinsky — COMMENTED** [review]
@ravicious 
Sorry for the delay I have missed this PR. 
I have left some suggestions but generally the flow looks good.

File: lib/teleterm/gateway/local_proxy_middleware.go
Comment (ravicious) [reply lib/teleterm/gateway/local_proxy_middleware.go:43]: > The CheckDBCerts seems to also returns diffrents errors apart from `x509.CertificateInvalidError` like: `certificate subject is for user` Should we call onExpireCert login in case of "certificate subject is for user" error ?

We shouldn't, I'll post a fix in a sec which inspects the error more closely.

File: lib/teleterm/gateway/local_proxy_middleware.go
Comment (ravicious) [inline lib/teleterm/gateway/local_proxy_middleware.go:54]: I don't know how unorthodox this is in Go, but I return early if there's no error or if the error is not related to cert expiry. This way there's less indentation around the code which does the actual work of calling `onExpiredCert`

File: lib/teleterm/gateway/local_proxy_middleware.go
Comment (ravicious) [inline lib/teleterm/gateway/local_proxy_middleware.go:59]: I didn't like how it showed up in the logs. `trace.NewAggregate` treats the first error as more important. But here we really care only about `onExpiredCertErr` - `err` is logged separately above anyway.

**Review by smallinsky — APPROVED** [review]

File: lib/teleterm/gateway/local_proxy_middleware.go
Comment (GavinFrazar) [reply lib/teleterm/gateway/local_proxy_middleware.go:54]: I like early return - less "trailing off to the right" in complicated logic

File: lib/teleterm/gateway/local_proxy_middleware.go
Comment (GavinFrazar) [reply lib/teleterm/gateway/local_proxy_middleware.go:45]: CheckDBCerts looks at db name/user in the database route in case user has previously logged in with `tsh db login` using a different db name/user than what they used for `tsh proxy db --tunnel --db-user=$dbuser --db-name=$dbname`, and the previous cert was loaded from disk

For teleterm flow, why should a database route mismatch not trigger cert reissue?

Comment (ravicious) [reply lib/teleterm/gateway/local_proxy_middleware.go:45]: Good question, I went back and forth on this when writing this reply.

In general, Connect doesn't let you execute the equivalent of `tsh db login` and `tsh proxy db --tunnel` separately. Both always happen within a single UI interaction of creating the gateway so you cannot provide different args to them. During gateway creation, we obtain a lock on `daemon.Service.mu`. This way we guarantee that no two gateways are being created at the same time… but only within Connect. The cert for the gateway is fetched right before we create it, but unlike `tsh proxy db` we save and load it from disk, even during reissue managed by the middleware.

And I guess here's where problems might happen. Technically we cannot guarantee that between being saved and loaded the cert is not modified by, let's say, `tsh db login`. Connect also has no real need to save the cert to disk, this is just the tech debt of not supporting per-session MFA: Connect calls `TeleportClient.ReissueUserCerts` (which doesn't return the cert) rather than `TeleportClient.IssueUserCertsWithMFA`.

This quarter we also plan to make Connect use `~/.tsh` rather than its own `TELEPORT_HOME` so the risk of some weird mismatched will be increased.

From what I see, the db cert on disk is scoped per cluster user & db server name, so I suppose we could run into something like this:

```mermaid
sequenceDiagram
    participant connect as Connect
    participant keys as key agent
    participant tsh as outside tsh
    connect ->> keys: Save db cert for db user foo
    tsh ->> keys: Save db cert for db user bar
    connect ->> keys: Get cert for db
    keys ->> connect: Return cert for db user bar
    connect ->> connect: Start local proxy
```

Which means:

* The mismatch between the cert subject and the local proxy params won't happen in Connect unless an incorrect cert was loaded into the gateway.
* For now, Connect cannot avoid loading certs from the disk.
* …so Connect should `CheckDBCerts` before instantiating `alpn.LocalProxy` and before doing `alpn.LocalProxy.SetCerts` when reissuing certs and return an error if there's a mismatch.
* By always setting the cert only after we check it, we ensure that in Connect, `OnNewConnection` cannot encounter a cert subject mismatch.
* …so Connect should treat that as unrecoverable error and not execute `OnExpiredCert`.

@GavinFrazar So I think the way to proceed would be to keep current code, that is Connect's middleware will not call `OnExpiredCert` if the check failed due to a subject mismatch. I'm going to extract `alpnproxy.CheckDBCerts` out of `alpnproxy.LocalProxy.CheckDBCerts` so that Connect can check the certs before actually setting them.

Does that sound good?

Comment (ravicious) [reply lib/teleterm/gateway/local_proxy_middleware.go:45]: Checking the certs will happen in the next PR where I actually utilize the callback, this one again is just about checking if they expired.

Comment (ravicious) [reply lib/teleterm/gateway/local_proxy_middleware.go:45]: This will be a bit of a pain though as I need to modify every test to create a valid cert for the gateway before we do `gateway.New`. 😩 I cannot simply do it inside `OnStart` in the middleware because from the user's perspective, it makes no sense to start a gateway with invalid certs. I'd rather fail during gateway creation and let the user reattempt the action, which will reobtain the certs.

Comment (ravicious) [reply lib/teleterm/gateway/local_proxy_middleware.go:45]: I'm going to do something else in the mean time in case this plan makes no sense, so I'd appreciate if you could take a look and let me know if I'm not overengineering this.

Comment (ravicious) [reply lib/teleterm/gateway/local_proxy_middleware.go:45]: But essentially this would entail fixing some tech debt / possible bug in Connect related to proxies that we simply weren't aware of.

Comment (ravicious) [reply lib/teleterm/gateway/local_proxy_middleware.go:45]: Sorry for spamming comments here. I've just realized that instead of executing the entirety of `CheckDBCerts` befor eConnect is creating a local proxy or before setting the certs, I can check for the subject mismatch alone. Essentially this is what we care about at that point (as I described in the comment with the graph above). In Connect it's not possible to start a proxy with an expired cert so the only concern would be having a cert that's for another username.

Comment (GavinFrazar) [reply lib/teleterm/gateway/local_proxy_middleware.go:45]: > By always setting the cert only after we check it, we ensure that in Connect, OnNewConnection cannot encounter a cert subject mismatch.
…so Connect should treat that as unrecoverable error and not execute OnExpiredCert.

I guess my thinking is that requesting new certs from the cluster is a way to handle that error - however you're saying it's a situation that shouldn't happen, and therefore it should be "unrecoverable". I assume you mean "unrecoverable" as in, the connection just gets dropped/closed instead of trying to handle the nonsensical error. If that's what you mean, then that's fine. However you want to handle the error is ok, I was just curious why the logic was different from tsh middleware.

**Review by GavinFrazar — APPROVED** [review]

File: lib/teleterm/gateway/local_proxy_middleware.go
Comment (GavinFrazar) [inline lib/teleterm/gateway/local_proxy_middleware.go:62]:
```suggestion
	return trace.Wrap(m.onExpiredCert(ctx))
```

nit

File: lib/teleterm/gateway/local_proxy_middleware.go
Comment (ravicious) [reply lib/teleterm/gateway/local_proxy_middleware.go:45]: > I assume you mean "unrecoverable" as in, the connection just gets dropped/closed instead of trying to handle the nonsensical error. If that's what you mean, then that's fine. 

Yep, exactly.

I'll have to reconsider this when adding per-session MFA support anyway, as I'd like to reuse a single middleware between tsh and Connect. But it doesn't seem like we'll touch per-session MFA in Connect in the next two or three months.

File: lib/teleterm/gateway/local_proxy_middleware_test.go
Comment (ravicious) [reply lib/teleterm/gateway/local_proxy_middleware_test.go:99]: > This is typically done as a way to avoid problems when running the table tests in parallel - did you mean to also add `t.Parallel()` inside `t.Run`?

Ah, no, just a leftover from copying parallel table tests from somewhere else. I don't think it's worth it to execute them in parallel, those are just three very simple tests.

**Pull request merged**
